apps/test-aeat-6012/src/main.cpp

```diff
@@ -1,127 +1,69 @@
+
+    updater_thread.start(&Updater);
+    reader_thread.start(&Reader);
+
+    updater_thread.join();
+    reader_thread.join();
+  } else {
+    while (true) {
+      MBED_ASSERT(encoder.update(callback));
+      ThisThread::sleep_for(PERIOD);
+    }
   }
 }
 
-void encoder_read_blocking(void) {
-  printf("\r\n--- AEAT-6012 Blocking Driver Test ---\r\n\r\n");
-  float measurement;
-  encoder.reset();
-
+void Updater() {
   while (true) {
-    timer.reset();
-    timer.start();
-
-    if (encoder.getAngleDeg(measurement)) {
-      timer.stop();
-
-      encoder_angle_deg        = measurement;
-      encoder_pos_read_time_us = std::chrono::duration_cast<std::chrono::microseconds>(timer.elapsed_time()).count();
-
-    } else {
-      printf("Encoder read FAILED!\r\n");
-    }
-
-    ThisThread::sleep_for(ENCODER_READ_PERIOD);
-
-    timer.reset();
+    Timer timer;
     timer.start();
-
-    if (encoder.getAngularVelocityDegPerSec(measurement)) {
-      timer.stop();
-
-      encoder_angular_vel_deg_per_sec = measurement;
-      encoder_vel_read_time_us = std::chrono::duration_cast<std::chrono::microseconds>(timer.elapsed_time()).count();
-
-    } else {
-      printf("Encoder read FAILED!\r\n");
-    }
-
-    ThisThread::sleep_for(ENCODER_READ_PERIOD);
+    MBED_ASSERT(encoder.update());
+    timer.stop();
+    std::string str;
+    str = "Time taken to update encoder: " + std::to_string(timer.elapsed_time().count()) + "us\r\n";
+    print(str);
+    ThisThread::sleep_for(PERIOD);
   }
 }
 
-void encoder_read_async(void) {
-  printf("\r\n--- AEAT-6012 Async Driver Test ---\r\n\r\n");
-  encoder.reset();
-
+void Reader() {
   while (true) {
-    timer.reset();
-    timer.start();
-
-    uint16_t attempts = 5;
-    while (attempts > 0 && !encoder.readAsync(callback)) {
-      attempts--;
-    }
-
-    if (attempts == 0) {
-      printf("ENCODER READ FAILED AFTER TOO MANY RETRIES\r\n");
-    } else {
-      // Wait for callback
-      while (!encoder_position_updated)
-        ;
-
-      timer.stop();
-
-      encoder_position_updated = false;
-
-      encoder_angle_deg               = encoder.getAngleDegNoTrigger();
-      encoder_angular_vel_deg_per_sec = encoder.getAngularVelocityDegPerSecNoTrigger();
-      encoder_pos_read_time_us        = encoder_vel_read_time_us =
-          std::chrono::duration_cast<std::chrono::microseconds>(timer.elapsed_time()).count();
-    }
-
-    ThisThread::sleep_for(ENCODER_READ_PERIOD);
+    std::string str;
+    str = "Angle: " + std::to_string(encoder.getAngleDeg()) +
+          ", Angular Velocity: " + std::to_string(encoder.getAngularVelocityDegPerSec()) + "\r\n";
+    print(str);
+    ThisThread::sleep_for(PERIOD);
   }
 }
 
-void callback(void) {
-  encoder_position_updated = true;
+void print(const std::string &str) {
+  std::unique_lock<Mutex> lock(mutex);
+  printf("%s", str.c_str());
 }
+
+void callback() {
+  std::string msg;
+  msg = "Angle: " + std::to_string(encoder.getAngleDeg()) +
```

Comment:
This might fail because getAngleDeg() locks a mutex

Reply:
i dont see why this would fail. the callback isnt called in an ISR.

Reply:
I'm pretty sure it will, the user callback gets called within the internal SPI transfer callback